src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,566 @@
+
+	PlannerRestrictionContext *plannerRestrictionContext =
+		FilterPlannerRestrictionForQuery(context->plannerRestrictionContext, query);
+	while (rteToSubqueryConverterReference && !IsRouterPlannable(query,
+																 plannerRestrictionContext))
+	{
+		ReplaceRTERelationWithRteSubquery(
+			rteToSubqueryConverterReference->rangeTableEntry,
+			rteToSubqueryConverterReference->restrictionList,
+			rteToSubqueryConverterReference->
+			requiredAttributeNumbers,
+			context);
+		RemoveFromRTEToSubqueryConverterContext(rteToSubqueryConverterContext,
+												rteToSubqueryConverterReference->
+												rangeTableEntry->relid);
+		rteToSubqueryConverterReference =
+			GetNextRTEToConvertToSubquery(query->jointree, rteToSubqueryConverterContext,
+										  context->plannerRestrictionContext,
+										  resultRelationId);
+	}
+}
+
+
+/*
+ * GetResultRelationId gets the result relation id from query
+ * if it exists.
+ */
+static Oid
+GetResultRelationId(Query *query)
+{
+	RangeTblEntry *resultRelation = ExtractResultRelationRTE(query);
+	Oid resultRelationId = InvalidOid;
+	if (resultRelation)
+	{
+		resultRelationId = resultRelation->relid;
+	}
+	return resultRelationId;
+}
+
+
+/*
+ * GetRangeTableEntriesFromJoinTree gets the range table entries that are
+ * on the given join tree.
+ */
+static void
+GetRangeTableEntriesFromJoinTree(Node *joinNode, List *rangeTableList,
+								 List **joinRangeTableEntries)
+{
+	if (joinNode == NULL)
+	{
+		return;
+	}
+	else if (IsA(joinNode, FromExpr))
+	{
+		FromExpr *fromExpr = (FromExpr *) joinNode;
+		Node *fromElement;
+
+		foreach_ptr(fromElement, fromExpr->fromlist)
+		{
+			GetRangeTableEntriesFromJoinTree(fromElement, rangeTableList,
+											 joinRangeTableEntries);
+		}
+	}
+	else if (IsA(joinNode, JoinExpr))
+	{
+		JoinExpr *joinExpr = (JoinExpr *) joinNode;
+		GetRangeTableEntriesFromJoinTree(joinExpr->larg, rangeTableList,
+										 joinRangeTableEntries);
+		GetRangeTableEntriesFromJoinTree(joinExpr->rarg, rangeTableList,
+										 joinRangeTableEntries);
+	}
+	else if (IsA(joinNode, RangeTblRef))
+	{
+		int rangeTableIndex = ((RangeTblRef *) joinNode)->rtindex;
+		RangeTblEntry *rte = rt_fetch(rangeTableIndex, rangeTableList);
+		*joinRangeTableEntries = lappend(*joinRangeTableEntries, rte);
+	}
+	else
+	{
+		pg_unreachable();
+	}
+}
+
+
+/*
+ * GetNextRTEToConvertToSubquery returns the range table entry
+ * which should be converted to a subquery. It considers the local join policy
+ * and result relation.
+ */
+static RTEToSubqueryConverterReference *
+GetNextRTEToConvertToSubquery(FromExpr *joinTree,
+							  RTEToSubqueryConverterContext *rteToSubqueryConverterContext,
+							  PlannerRestrictionContext *plannerRestrictionContext, Oid
+							  resultRelationId)
+{
+	RTEToSubqueryConverterReference *localRTECandidate = NULL;
+	RTEToSubqueryConverterReference *distributedRTECandidate = NULL;
+	if (!FillLocalAndDistributedRTECandidates(rteToSubqueryConverterContext,
```

Comment:
Now I am thinking of calling `ContainsTableToBeConvertedToSubquery` here to decide whether we should continue or not, after all if three is no table to be converted to subquery we shouldn't continue. This might simplify things if it works because that also means we have a single place to decide that

Reply:
Actually I am simplifying this even further by using `ShouldConvertLocalTableJoinsToSubqueries`, it seems to be working.